slkspec/core.py

```diff
@@ -1,129 +1,355 @@
+    in any kind of write mode."""
+
+    def __init__(
+        self,
+        url: str,
+        local_file: str,
+        *,
+        override: bool = True,
+        mode: str = "rb",
+        touch: bool = True,
+        file_permissions: int = 0o3777,
+        _lock: threading.Lock = _retrieval_lock,
+        _file_queue: Queue[Tuple[str, str]] = FileQueue,
+        **kwargs: Any,
+    ):
+        if not set(mode) & set("r"):  # The mode must have a r
+            raise NotImplementedError(self.write_msg)
+        if "b" not in mode:
+            kwargs.setdefault("encoding", "utf-8")
+        self._file = str(Path(local_file).expanduser().absolute())
+        self._url = str(url)
+        self.touch = touch
+        self.file_permissions = file_permissions
+        self._order_num = 0
+        self._file_obj: Optional[IO[Any]] = None
+        self._lock = _lock
+        self.kwargs = kwargs
+        self.mode = mode
+        self.newlines = None
+        self.error = "strict"
+        self.encoding = kwargs.get("encoding")
+        self.write_through = False
+        self._file_queue = _file_queue
+        print(self._file)
+        with _lock:
+            if not Path(self._file).exists() or override:
+                self._file_queue.put((self._url, str(Path(self._file).parent)))
+            elif Path(self._file).exists():
+                if self.touch:
+                    Path(self._file).touch()
+                self._file_obj = open(self._file, mode, **kwargs)
+
+    @property
+    def name(self) -> str:
+        """Get the file for the SLKFile object."""
+        if self._file_obj is not None:
+            return self._file
+        return self._url
+
+    def _retrieve_items(self, retrieve_files: list[tuple[str, str]]) -> None:
+        """Get items from the tape archive."""
+
+        retrieval_requests: Dict[Path, List[str]] = defaultdict(list)
+        logger.critical("Retrieving %i items from tape", len(retrieve_files))
+        for inp_file, out_dir in retrieve_files:
+            retrieval_requests[Path(out_dir)].append(inp_file)
+        for output_dir, inp_files in retrieval_requests.items():
+            output_dir.mkdir(parents=True, exist_ok=True, mode=self.file_permissions)
+            logger.critical("Creating slk query for %i files", len(inp_files))
+            search_str = pyslk.slk_search(pyslk.slk_gen_file_query(inp_files))
+            search_id_re = re.search("Search ID: [0-9]*", search_str)
+            if search_id_re is None:
+                raise ValueError("No files found in archive.")
+            search_id = int(search_id_re.group(0)[11:])
+            logger.critical("Retrieving files for search id: %i", search_id)
+            pyslk.slk_retrieve(search_id, str(output_dir))
+            logger.critical("Adjusting file permissions")
+            for out_file in map(Path, inp_files):
+                out_file.chmod(self.file_permissions)
```

Comment:
`out_file` is the path on tape if I see it correctly. The `slk_cache` needs to be added.

# MRE
```python
import os
import fsspec
os.environ["SLK_CACHE"] = "/scratch/m/<USER>/retrieval/"
m = fsspec.open("slk:///arch/mh0010/m300408/experiments/EUREC4A/control/mfr/20200131T22/INDEX.txt")
with m as f:
     for i in range(10):
        print(f.readline())
```

<details>
<summary>FileNotFoundError</summary>
<pre>
Input In [2], in <cell line: 3>()
      3 with m as f:
      4     for i in range(10):
----> 5         print(f.readline())

File ~/GitProjects/slkspec_dev/slkspec/core.py:208, in SLKFile.read(self, size)
    201 """The the content of a file-stream.
    202 
    203 size: int, default: -1
    204     read at most size characters from the stream, -1 means everything
    205     is read.
    206 """
    207 if self._file_obj is None:
--> 208     self._cache_files()
    209 return self._file_obj.read(size)

File ~/GitProjects/slkspec_dev/slkspec/core.py:162, in SLKFile._cache_files(self)
    160     items.append(self._file_queue.get())
    161     self._file_queue.task_done()
--> 162 self._retrieve_items(items)
    163 _ = self._file_queue.get()
    164 self._file_queue.task_done()

File ~/GitProjects/slkspec_dev/slkspec/core.py:152, in SLKFile._retrieve_items(self, retrieve_files)
    150 logger.critical("Adjusting file permissions")
    151 for out_file in map(Path, inp_files):
--> 152     out_file.chmod(self.file_permissions)

File ~/.conda/envs/slkspec_dev/lib/python3.10/pathlib.py:1191, in Path.chmod(self, mode, follow_symlinks)
   1187 def chmod(self, mode, *, follow_symlinks=True):
   1188     """
   1189     Change the permissions of the path, like os.chmod().
   1190     """
-> 1191     self._accessor.chmod(self, mode, follow_symlinks=follow_symlinks)

FileNotFoundError: [Errno 2] No such file or directory: '/arch/mh0010/m300408/experiments/EUREC4A/control/mfr/20200131T22/INDEX.txt'
</pre>
</details>

Reply:
Yes you're right, I've just figured the same.

Reply:
Btw thanks for the exmaple

Reply:
Ok that worked now:

```python
url = fsspec.open("slk:///arch/mh0010/m300408/experiments/EUREC4A/control/mfr/20200131T22/INDEX.txt")
/home/b/b380001/freva/lib/python3.10/site-packages/fsspec/spec.py:76: UserWarning: The slk_cache variable nor the SLK_CACHE environmentvariable wasn't set. Falling back to default /scratch/k/k204230
  obj = super().__call__(*args, **kwargs)

 with url as f:
   for i in range(10):
       print(f.readline())
    
/scratch/k/k204230/arch/mh0010/m300408/experiments/EUREC4A/control/mfr/20200131T22/INDEX.txt
slk search '{"$and":[{"path":{"$gte":"/arch/mh0010/m300408/experiments/EUREC4A/control/mfr/20200131T22","$max_depth":1}},{"resources.name":{"$regex":"INDEX.txt"}}]}'
b'/work/mh0010/m300408/DVC-test/EUREC4A-ICON/EUREC4A/run\n'
b'command_line: packems -j 12 -o eurec4a_restart_20200131T22 -S /arch/mh0010/m300408/experiments/EUREC4A/control/mfr/20200131T22 -O by_time /work/mh0010/m300408/DVC-test/EUREC4A-ICON/EUREC4A/experiments/EUREC4A/multifile_restart_atm_20200131T220000Z.mfr/\n'
b'version: 2.0.2\n'
b'---\n'
b'-rw-r--r-- m300408/mh0010 53998 2022-04-16 01:40 /work/mh0010/m300408/DVC-test/EUREC4A-ICON/EUREC4A/experiments/EUREC4A/multifile_restart_atm_20200131T220000Z.mfr/attributes.nc -> eurec4a_restart_20200131T22_005.tar\n'
b'-rw-r--r-- m300408/mh0010 17289274 2022-04-16 01:40 /work/mh0010/m300408/DVC-test/EUREC4A-ICON/EUREC4A/experiments/EUREC4A/multifile_restart_atm_20200131T220000Z.mfr/patch1_0.nc -> eurec4a_restart_20200131T22_005.tar\n'
b'-rw-r--r-- m300408/mh0010 14717089 2022-04-16 01:40 /work/mh0010/m300408/DVC-test/EUREC4A-ICON/EUREC4A/experiments/EUREC4A/multifile_restart_atm_20200131T220000Z.mfr/patch1_1.nc -> eurec4a_restart_20200131T22_005.tar\n'
b'-rw-r--r-- m300408/mh0010 13355913 2022-04-16 01:40 /work/mh0010/m300408/DVC-test/EUREC4A-ICON/EUREC4A/experiments/EUREC4A/multifile_restart_atm_20200131T220000Z.mfr/patch1_10.nc -> eurec4a_restart_20200131T22_005.tar\n'
b'-rw-r--r-- m300408/mh0010 13430198 2022-04-16 01:42 /work/mh0010/m300408/DVC-test/EUREC4A-ICON/EUREC4A/experiments/EUREC4A/multifile_restart_atm_20200131T220000Z.mfr/patch1_100.nc -> eurec4a_restart_20200131T22_005.tar\n'
b'-rw-r--r-- m300408/mh0010 13233198 2022-04-16 01:40 /work/mh0010/m300408/DVC-test/EUREC4A-ICON/EUREC4A/experiments/EUREC4A/multifile_restart_atm_20200131T220000Z.mfr/patch1_1000.nc -> eurec4a_restart_20200131T22_001.tar\n'
```